File: 1.py
# Machine Opcode Table (MOT)
MOT = {
    "MOVER": {"type": "IS", "code": 1},
    "ADD": {"type": "IS", "code": 2},
    "MOVEM": {"type": "IS", "code": 3},
    "START": {"type": "AD", "code": 1},
    "END": {"type": "AD", "code": 2},
    "DS": {"type": "DL", "code": 1},
    "DC": {"type": "DL", "code": 2},
}

# Register Table
REGISTER_TABLE = {"AREG": 1, "BREG": 2, "CREG": 3, "DREG": 4}

# Initialize tables
symbol_table = {}
intermediate_code = []

# Input ALP Code
alp_code = [
    "START 100",
    "MOVER BREG, C",
    "C DC 1",
    "D DS 2",
    "END",
]

# Process ALP Code
lc = 0  # Location Counter

for line in alp_code:
    tokens = line.split()
    if tokens[0] in MOT:  # First token is an instruction or directive
        entry = MOT[tokens[0]]
        if entry["type"] == "AD":  # Assembler Directive
            if tokens[0] == "START":
                lc = int(tokens[1])  # Initialize location counter
                intermediate_code.append((lc, f"AD {entry['code']} {tokens[1]}"))
            elif tokens[0] == "END":
                intermediate_code.append((lc, f"AD {entry['code']}"))
        elif entry["type"] == "IS":  # Imperative Statement
            opcode = entry["code"]
            reg = tokens[1][:-1]  # Remove ',' from the register
            operand = tokens[2]  # Symbol (e.g., C)
            reg_code = REGISTER_TABLE.get(reg, None)
            if operand not in symbol_table:
                symbol_table[operand] = None  # Forward reference
            intermediate_code.append((lc, f"IS {opcode} {reg_code} {operand}"))
            lc += 1
        elif entry["type"] == "DL":  # Declarative Statement
            if tokens[0] == "DC":  # Define constant
                symbol_table[tokens[1]] = lc
                intermediate_code.append((lc, f"DL {entry['code']} {tokens[2]}"))
                lc += 1
            elif tokens[0] == "DS":  # Define storage
                symbol_table[tokens[1]] = lc
                intermediate_code.append((lc, f"DL {entry['code']} {tokens[2]}"))
                lc += int(tokens[2])  # Reserve space
    else:  # First token is a label (e.g., C or D)
        symbol_table[tokens[0]] = lc
        if tokens[1] == "DC":  # Handle DC after label
            entry = MOT[tokens[1]]
            intermediate_code.append((lc, f"DL {entry['code']} {tokens[2]}"))
            lc += 1
        elif tokens[1] == "DS":  # Handle DS after label
            entry = MOT[tokens[1]]
            intermediate_code.append((lc, f"DL {entry['code']} {tokens[2]}"))
            lc += int(tokens[2])

# Display Output
print("\nInput ALP Code:")
for line in alp_code:
    print(line)

print("\nSymbol Table:")
for symbol, address in symbol_table.items():
    print(f"{symbol} -> {address}")

print("\nIntermediate Code:")
for entry in intermediate_code:
    print(entry)

File: 1.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_LINES 100
#define MAX_SYMBOLS 100

// Machine Opcode Table (MOT)
typedef struct {
    char instruction[10];
    char type[3];
    int code;
} MOTEntry;

MOTEntry MOT[] = {
    {"MOVER", "IS", 1},
    {"ADD", "IS", 2},
    {"MOVEM", "IS", 3},
    {"START", "AD", 1},
    {"END", "AD", 2},
    {"DS", "DL", 1},
    {"DC", "DL", 2}
};

// Register Table
typedef struct {
    char reg[10];
    int code;
} RegisterEntry;

RegisterEntry REGISTER_TABLE[] = {
    {"AREG", 1},
    {"BREG", 2},
    {"CREG", 3},
    {"DREG", 4}
};

// Symbol Table
typedef struct {
    char symbol[10];
    int address;
} Symbol;

Symbol symbol_table[MAX_SYMBOLS];
int symbol_count = 0;

// Intermediate Code Entry
typedef struct {
    int location;
    char code[50];
} IntermediateCode;

IntermediateCode intermediate_code[MAX_LINES];
int intermediate_count = 0;

// Helper Functions
int searchMOT(char *instruction) {
    for (int i = 0; i < sizeof(MOT) / sizeof(MOTEntry); i++) {
        if (strcmp(MOT[i].instruction, instruction) == 0) {
            return i;
        }
    }
    return -1;
}

int searchRegisterTable(char *reg) {
    for (int i = 0; i < sizeof(REGISTER_TABLE) / sizeof(RegisterEntry); i++) {
        if (strcmp(REGISTER_TABLE[i].reg, reg) == 0) {
            return REGISTER_TABLE[i].code;
        }
    }
    return -1;
}

int searchSymbolTable(char *symbol) {
    for (int i = 0; i < symbol_count; i++) {
        if (strcmp(symbol_table[i].symbol, symbol) == 0) {
            return i;
        }
    }
    return -1;
}

void addToSymbolTable(char *symbol, int address) {
    strcpy(symbol_table[symbol_count].symbol, symbol);
    symbol_table[symbol_count].address = address;
    symbol_count++;
}

void addIntermediateCode(int location, char *code) {
    intermediate_code[intermediate_count].location = location;
    strcpy(intermediate_code[intermediate_count].code, code);
    intermediate_count++;
}

// Main Function
int main() {
    char *alp_code[] = {
        "START 100",
        "MOVER BREG, C",
        "C DC 1",
        "D DS 2",
        "END"
    };
    int line_count = 5;
    int lc = 0; // Location Counter

    for (int i = 0; i < line_count; i++) {
        char line[50];
        strcpy(line, alp_code[i]);
        char *tokens[5];
        int token_count = 0;

        char *token = strtok(line, " ,");
        while (token) {
            tokens[token_count++] = token;
            token = strtok(NULL, " ,");
        }

        int motIndex = searchMOT(tokens[0]);
        if (motIndex != -1) {
            MOTEntry entry = MOT[motIndex];
            if (strcmp(entry.type, "AD") == 0) {
                if (strcmp(tokens[0], "START") == 0) {
                    lc = atoi(tokens[1]);
                    char code[50];
                    sprintf(code, "AD %d %s", entry.code, tokens[1]);
                    addIntermediateCode(lc, code);
                } else if (strcmp(tokens[0], "END") == 0) {
                    char code[50];
                    sprintf(code, "AD %d", entry.code);
                    addIntermediateCode(lc, code);
                }
            } else if (strcmp(entry.type, "IS") == 0) {
                int reg_code = searchRegisterTable(tokens[1]);
                if (searchSymbolTable(tokens[2]) == -1) {
                    addToSymbolTable(tokens[2], -1);
                }
                char code[50];
                sprintf(code, "IS %d %d %s", entry.code, reg_code, tokens[2]);
                addIntermediateCode(lc, code);
                lc++;
            } else if (strcmp(entry.type, "DL") == 0) {
                addToSymbolTable(tokens[1], lc);
                char code[50];
                sprintf(code, "DL %d %s", entry.code, tokens[2]);
                addIntermediateCode(lc, code);
                lc += (strcmp(tokens[0], "DS") == 0) ? atoi(tokens[2]) : 1;
            }
        } else {
            addToSymbolTable(tokens[0], lc);
            int motIndexLabel = searchMOT(tokens[1]);
            if (motIndexLabel != -1) {
                MOTEntry labelEntry = MOT[motIndexLabel];
                char code[50];
                sprintf(code, "DL %d %s", labelEntry.code, tokens[2]);
                addIntermediateCode(lc, code);
                lc += (strcmp(tokens[1], "DS") == 0) ? atoi(tokens[2]) : 1;
            }
        }
    }

    printf("\nInput ALP Code:\n");
    for (int i = 0; i < line_count; i++) {
        printf("%s\n", alp_code[i]);
    }

    printf("\nSymbol Table:\n");
    for (int i = 0; i < symbol_count; i++) {
        printf("%s -> %d\n", symbol_table[i].symbol, symbol_table[i].address);
    }

    printf("\nIntermediate Code:\n");
    for (int i = 0; i < intermediate_count; i++) {
        printf("%d %s\n", intermediate_code[i].location, intermediate_code[i].code);
    }

    return 0;
}
